src/skin.cpp

```diff
@@ -545,6 +546,31 @@ Polygons SkinInfillAreaComputation::generateNoAirAbove(SliceLayerPart& part)
     return no_air_above;
 }
 
+/*
+ * This function is executed in a parallel region based on layer_nr.
+ * When modifying make sure any changes does not introduce data races.
+ *
+ * this function may only read the skin and infill from the *current* layer.
+ */
+    Polygons SkinInfillAreaComputation::generateNoAirBelow(SliceLayerPart& part, size_t flooring_layer_count)
+    {
+        const size_t wall_idx = std::min(size_t(2), mesh.settings.get<size_t>("wall_line_count"));
+        if (layer_nr - flooring_layer_count >= 0)
+        {
+            Polygons no_air_below = getWalls(part, layer_nr - flooring_layer_count, wall_idx);
+            if (!no_small_gaps_heuristic)
+            {
+                for (int layer_nr_below = layer_nr - flooring_layer_count + 1; layer_nr_below < layer_nr; layer_nr_below++)
+                {
+                    Polygons outlines_below = getWalls(part, layer_nr_below, wall_idx);
+                    no_air_below = no_air_below.intersection(outlines_below);
+                }
+            }
+            return no_air_below;
+        }
+        return {};
```

Comment:
Store `layer_nr - flooring_layer_count` in variable for reuse.
Don't do calculations as `for (int layer_nr_below = layer_nr - flooring_layer_count + 1;` in a for loop, most modern compilers should be able to make that distinction but in complex files/functions they can make mistakes and do the calculation for each pass.

The previous early out ensures that the subtractions never result in `< 0` so we can keep on using `size_t` removing implicit casts.

> **NOTE**  
> Untested code below

```suggestion
       const size_t non_flooring_layer = layer_nr - flooring_layer_count;
       Polygons no_air_below = getWalls(part, non_flooring_layer, wall_idx);
       if (!no_small_gaps_heuristic)
       {
           const size_t next_non_flooring_layer = non_flooring_layer + 1;
           for (size_t layer_nr_below = next_non_flooring_layer; layer_nr_below < layer_nr; layer_nr_below++)
           {
               Polygons outlines_below = getWalls(part, layer_nr_below, wall_idx);
               no_air_below = no_air_below.intersection(outlines_below);
           }
       }
       return no_air_below;
```